### PR TITLE
Make ts-batch work for oracle

### PR DESCRIPTION
**Pull request by forus**
Please also review https://github.com/transmart/transmart-data/pull/118

### CODE REVIEW

File: src/main/groovy/org/transmartproject/batch/batchartifacts/DbMetadataBasedBoundsValidator.groovy [hunk 1]
```diff
@@ -83,14 +82,24 @@ class DbMetadataBasedBoundsValidator implements Validator {
...
-        // TODO: check which character oracle uses to escape the like expressions
+
         ResultSet rs = meta.getColumns(
                 null,
-                StringUtils.escapeForLike(spec.schema),
-                StringUtils.escapeForLike(spec.table),
-                StringUtils.escapeForLike(spec.column),)
+                getIdentifier(meta, spec.schema),
+                getIdentifier(meta, spec.table),
+                getIdentifier(meta, spec.column))
```
Comment (cataphract) [inline src/main/groovy/org/transmartproject/batch/batchartifacts/DbMetadataBasedBoundsValidator.groovy]: The javadocs say this is a `LIKE` expression. Therefore, the `_` and `%` must be escaped.

File: src/main/groovy/org/transmartproject/batch/concept/ConceptTree.groovy [hunk 2]
```diff
@@ -40,24 +42,22 @@ class ConceptTree {
...
 
             nodeMap[n.path] = n
+            savedNodes << n
         }
     }
 
-    Collection<ConceptNode> getNewConceptNodes() {
-        nodeMap.values().findAll { it.isNew() }
+    Set<ConceptNode> getNewConceptNodes() {
+        nodeMap.values().toSet() - savedNodes
```
Comment (cataphract) [inline src/main/groovy/org/transmartproject/batch/concept/ConceptTree.groovy]: Avoid using the `-` operator on sets. It's very inefficient. 

    groovy -e '(1..50000 as Set) - (1..20000 as Set)'`

Takes 9 seconds on my machine, whereas:

    CLASSPATH=~/.m2/repository/com/google/guava/guava/18.0/guava-18.0.jar \
        time groovy -e 'println com.google.common.collect.Sets.difference((1..50000 as Set),(1..20000 as Set))'

Takes less than 1 (most of which is JVM startup).

File: src/main/groovy/org/transmartproject/batch/concept/ConceptTree.groovy [hunk 3]
```diff
@@ -112,14 +112,18 @@ class ConceptTree {
...
     void reserveIdsFor(ConceptNode node) {
-        if (!node.isNew()) {
+        if (isSavedNode(node)) {
             log.warn("Could not rewrite id for node that already has one (${node.code}).")
             return
         }
 
         node.code = reserver.getNext(Sequences.CONCEPT)
-        node.i2b2RecordId = reserver.getNext(Sequences.I2B2_RECORDID)
+        savedNodes << node
```
Comment (cataphract) [inline src/main/groovy/org/transmartproject/batch/concept/ConceptTree.groovy]: This doesn't make sense. By reserving an id you're not saving any node.

File: src/main/groovy/org/transmartproject/batch/batchartifacts/DbMetadataBasedBoundsValidator.groovy (hunk 1 above)
Comment (forus) [inline src/main/groovy/org/transmartproject/batch/batchartifacts/DbMetadataBasedBoundsValidator.groovy]: @cataphract Do you have any idea how you could make it work for oracle? It seems there is no default escape symbol.

Comment (cataphract) [inline src/main/groovy/org/transmartproject/batch/batchartifacts/DbMetadataBasedBoundsValidator.groovy]: Looking at the code in the Oracle driver, it's using `/` as an escape character: `AND t.table_name LIKE :2 ESCAPE '/'\n  AND t.column_name LIKE :3 ESCAPE '/'`

Comment (cataphract) [inline src/main/groovy/org/transmartproject/batch/batchartifacts/DbMetadataBasedBoundsValidator.groovy]: And for Postgres it's the default `\`.

**Pull request merged**